Core/Core/Database/CoreData/Models/Permissions.swift

```diff
@@ -0,0 +1,173 @@
+    @NSManaged public var manageUserObservers: Bool
+    @NSManaged public var readCourseContent: Bool
+    @NSManaged public var readCourseList: Bool
+    @NSManaged public var readMessages: Bool
+    @NSManaged public var resetAnyMfa: Bool
+    @NSManaged public var siteAdmin: Bool
+    @NSManaged public var viewCourseChanges: Bool
+    @NSManaged public var viewErrorReports: Bool
+    @NSManaged public var viewGradeChanges: Bool
+    @NSManaged public var viewJobs: Bool
+    @NSManaged public var viewNotifications: Bool
+    @NSManaged public var viewStatistics: Bool
+    @NSManaged public var undeleteCourses: Bool
+
+    // Account and Course level
+    @NSManaged public var changeCourseState: Bool
+    @NSManaged public var commentOnOthersSubmissions: Bool
+    @NSManaged public var createCollaborations: Bool
+    @NSManaged public var createConferences: Bool
+    @NSManaged public var createForum: Bool
+    @NSManaged public var generateObserverPairingCode: Bool
+    @NSManaged public var importOutcomes: Bool
+    @NSManaged public var ltiAddEdit: Bool
+    @NSManaged public var manageAdminUsers: Bool
+    @NSManaged public var manageAssignments: Bool
+    @NSManaged public var manageCalendar: Bool
+    @NSManaged public var manageContent: Bool
+    @NSManaged public var manageFiles: Bool
+    @NSManaged public var manageGrades: Bool
+    @NSManaged public var manageGroups: Bool
+    @NSManaged public var manageInteractionAlerts: Bool
+    @NSManaged public var manageOutcomes: Bool
+    @NSManaged public var manageSections: Bool
+    @NSManaged public var manageStudents: Bool
+    @NSManaged public var manageUserNotes: Bool
+    @NSManaged public var manageRubrics: Bool
+    @NSManaged public var manageWiki: Bool
+    @NSManaged public var moderateForum: Bool
+    @NSManaged public var postToForum: Bool
+    @NSManaged public var readAnnouncements: Bool
+    @NSManaged public var readEmailAddresses: Bool
+    @NSManaged public var readForum: Bool
+    @NSManaged public var readQuestionBanks: Bool
+    @NSManaged public var readReports: Bool
+    @NSManaged public var readRoster: Bool
+    @NSManaged public var readSis: Bool
+    @NSManaged public var selectFinalGrade: Bool
+    @NSManaged public var sendMessages: Bool
+    @NSManaged public var sendMessagesAll: Bool
+    @NSManaged public var viewAllGrades: Bool
+    @NSManaged public var viewAuditTrail: Bool
+    @NSManaged public var viewGroupPages: Bool
+    @NSManaged public var viewUserLogins: Bool
+
+    @discardableResult
+    public static func save(_ item: APIPermissions, for context: Context, in client: PersistenceClient) throws -> Permissions {
+        let predicate = NSPredicate(format: "%K == %@", #keyPath(Permissions.context), context.canvasContextID)
+        let model: Permissions = client.fetch(predicate).first ?? client.insert()
+        model.context = context.canvasContextID
+        model.becomeUser = item.become_user ?? model.becomeUser
```

Comment:
The way we request these, we only want to change the ones we got back, right? If it doesn't come back, we can't assume it's `false`. That's what this code looks like it's doing. Did I understand right?

Reply:
Correct. If we previously get all of the permissions for the user but then only request one permission later, we don't want to overwrite all of the existing permissions. If we have no permissions and only request one, we want to set the rest as false until we know from a subsequent request that they are true.

Reply:
I think I misread your comment the first time in regards to assuming the permission is false. I initially went the route of marking the permissions as `Bool?` but that is not a possible type for an `@NSManaged` property. So I either had to use `NSBool` or make them false by default. I think it's safe to mark them as false by default. I'd rather that than having them be true by default. I would expect a screen that needs to know about a permission to request its true value on load. Does that seem reasonable?

Reply:
Yeah, I think that's reasonable.